### ISSUE TITLE
Host (CPU) RAM grows linearly during state-based GPU PPO training (stock StackCube-v1) → eventual OOM

### ISSUE THREAD

**Issue by jiayingtu21-hash** [haosulab/ManiSkill#1459]
## Describe the bug

When training PPO with GPU simulation and **state** observations using the official `examples/baselines/ppo/ppo.py`, **host (system) RAM grows roughly linearly with training steps and never plateaus**, eventually exhausting RAM (and swap) and causing an OOM / swap-thrash hang on a 16 GB machine.

It reproduces with a **stock task (`StackCube-v1`) and the unmodified official `ppo.py`**, so it is not specific to custom tasks or domain randomization. (A custom per-env DR task that builds 1024 separate actors leaks faster — more rigid bodies / more per-step state read — which is consistent with, but not proof of, the growth scaling with the amount of per-step state accessed.)

## Environment

- ManiSkill **3.0.1**, SAPIEN **3.0.3**
- Python 3.11, state-based PPO (`examples/baselines/ppo/ppo.py`)
- GPU: RTX 4070 Ti SUPER (16 GB); driver 595.71.05 / CUDA 13.2
- System RAM: 16 GB + 2 GB swap; Ubuntu

## To reproduce

1. `cd examples/baselines/ppo`
2. Run the official PPO on a **stock** task (eval disabled here only to isolate; the leak is present with or without eval):
   ```bash
   python ppo.py --env_id=StackCube-v1 --num_envs=1024 --update_epochs=8 \
     --num_minibatches=32 --total_timesteps=20000000 --eval_freq=100000 --no-capture-video
   ```
3. In parallel, watch host RAM (`watch -n5 free -h`, or sample `/proc/self/status` `VmRSS`).
4. Observe host RAM climbing ~linearly and not plateauing.

## Observed data

Stock `StackCube-v1`, `num_envs=1024`, eval disabled — process RSS over time (after the initial build/warm-up phase):

| t (min) | RSS (GB) |
|--------:|---------:|
| 6  | 8.50 |
| 8  | 8.59 |
| 10 | 8.63 |
| 12 | 8.67 |
| 14 | 8.73 |
| 16 | 8.83 |

≈ **+0.033 GB/min, roughly linear** (a custom per-env DR task on the same machine grows ≈ +0.055 GB/min). Extrapolated, this exhausts 16 GB and OOM/thrashes during a long run.

`tracemalloc` (started at process launch, snapshotting every 2 min) shows the **largest Python-visible growth sites are consistent across every snapshot and are all SAPIEN per-step state accessors** — note these account for only a *small fraction* of the total RSS growth (see the note below the list):

```
  +~1900 KB  raw_pose = self.px.cuda_rigid_body_data.torch()[
  +~1060 KB  raw_pose = self.px.cuda_rigid_body_data.torch()[self._body_data_index, :7]
  + ~760 KB  return getattr(self.px, self._body_data_name).torch()
  + ~710 KB  return self.px.cuda_articulation_qpos.torch()[
  + ~350 KB  pairwise_contact_impulses = query.cuda_impulses.torch().clone()
  + ~350 KB  self.px.cuda_articulation_target_qpos.torch()[
  + ~180 KB  return self.px.cuda_articulation_qvel.torch()[
```

The Python-visible growth above is only a small fraction of the total RSS growth, so most of the increase appears to be the underlying C/CUDA buffers these `.torch()` views reference.

## What I have isolated (and what I have not)

- **Bare env loop** (`env.step` + `env.reset`, no policy / no training): RSS **flat** over 2M+ steps — even though this calls the same state accessors.
- **Forcing `reconfiguration_freq=1`** and resetting every step (rebuilding the scene each reset): RSS **flat**.
- **Disabling eval** (`--eval_freq` huge) vs **frequent eval** (`--eval_freq=5`): **no difference** in growth rate.

So the growth **requires the full training loop, not bare stepping**. Because a bare loop that calls the same accessors stays flat, the accessors themselves are not the leak; rather, something in the training loop appears to **retain references** to the per-step state tensors (the autograd graph / rollout buffer are the obvious suspects), preventing their underlying buffers from being released. **I have not confirmed which layer retains them, so I cannot yet say whether the fix belongs in `ppo.py` or upstream in SAPIEN's tensor lifecycle.** It is **not** tied to reset/reconfigure (cf. #965) **nor** to eval.

## Expected behavior

Host RAM stable across training steps (no unbounded growth).

## Actual behavior

Host RAM grows ~linearly with steps until OOM / swap-thrash.

## Profiling script used (wraps the official `ppo.py`)

```python
import sys, os, time, threading, tracemalloc
tracemalloc.start(10)
import mani_skill  # only register stock tasks

def rss_gb():
    with open('/proc/self/status') as f:
        for line in f:
            if line.startswith('VmRSS'):
                return int(line.split()[1]) / 1048576.0
    return -1.0

def monitor():
    prev = None; n = 0
    while True:
        time.sleep(120); n += 1
        snap = tracemalloc.take_snapshot()
        print('\n== snapshot %d  t=%dmin  RSS=%.2fG ==' % (n, n*2, rss_gb()), flush=True)
        if prev is not None:
            for s in sorted(snap.compare_to(prev, 'lineno'),
                            key=lambda s: s.size_diff, reverse=True)[:12]:
                print('  %+8.1f KB  %s' % (s.size_diff/1024.0,
                      s.traceback.format()[-1].strip()), flush=True)
        prev = snap

threading.Thread(target=monitor, daemon=True).start()
sys.argv = ['ppo.py', '--env_id=StackCube-v1', '--num_envs=1024',
            '--update_epochs=8', '--num_minibatches=32',
            '--total_timesteps=20000000', '--eval_freq=100000', '--no-capture-video']
import runpy
runpy.run_path('ppo.py', run_name='__main__')
```

Happy to provide full logs, test other versions, or run additional isolation experiments.
